Staex Internet of Data

lavrd opened the pull request:
#### Project Abstract

Staex Internet of Data.

The goal of this project is to create a Web3 IoT data infrastructure with a stable economy. In simple words we want to simplify Web3 onboarding process and bring more IoT device owners to share their useful data with some profit and other people to find and use such data transparently, securely and easily for their life or research.

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [ ] **Level 2**:  Up to $30,000, 3 approvals
- [x] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (bank details via email _or_ Polkadot (USDC & USDT) address in the application).
- [x] I am aware that, in order to receive a grant, I (and the entity I represent) have to successfully complete a KYC/KYB check.
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

lavrd commented:
I have read and hereby sign the Contributor License Agreement.

Noc2 reviewed (changes requested):
Thanks a lot for the application. We usually prefer smart contract languages that are native to our ecosystem when it comes to grants. Would you be willing to change solidity to ink!? Also, are you still planning to apply for the https://futures.web3.foundation/? Potentially this would be a better fit.

lavrd commented:
Thanks for review! I have update application to use ink!, not a problem for us and overall looks better. What the difference between this program and [this](https://futures.web3.foundation/)? I checked template and it looks quite the same. Can we continue with this repository one and use [this](https://futures.web3.foundation/) for future work?

keeganquigley commented on applications/staex-iod.md:
```diff
@@ -0,0 +1,326 @@
+
+To give users possibility to find particular type of data or particular IoT device we need to index them somehow. Indexing part we want to implement through DIDs and Subsquid project.
+
+So we will implement ink! smart contract which will be deployed on Kusama network. This smart contract will have several methods:
```
Kusama isn't capable of running smart contracts since it is a relay chain. Did you mean to list a different chain or testnet here?

lavrd replied:
Yes, it was a mistake from me. While this phase of the project, maybe, it is not necessary to deploy smart contracts on testnet and for sure on mainnet. But for future it seems that we need to deploy them to Shiden parachain or to Shibuya while testing.

keeganquigley commented on applications/staex-iod.md:
```diff
@@ -0,0 +1,326 @@
+
+And here is a flow: 
+
+1. User puts some tokens to smart contract for exact owner and point which type of data it needs.
```
How would the user send tokens to the smart contract? Would they need to use XCM first to teleport the tokens to the chain that is hosting said contracts?

lavrd replied:
I think in next project phases we will implement this part like this:
- First time users will need to use UI to buy data access with particular tokens like SDN/SBY.
- In future for mass adoption we can implement a feature for UI to sell/buy data access using different tokens from different chains. For example IoT device owner wants to sell data for USDT, but a person wants to buy an access using DOT. Or SDN <-> KSM. So we can implement in UI auto teleport/exchange for user. As I understand it is possible through XCM and Polkadot.js.

keeganquigley commented on applications/staex-iod.md:
```diff
@@ -0,0 +1,326 @@
+- Requested data is not available at the moment.
+- Device has high amount of request (highload).
+
+This contract will use DID smart contract to check that user sent required amount of tokens otherwise reject automatically.
```
Would you be using an existing solution here, and if so, which one? Or would you be creating your own custom DID contract?

lavrd replied:
We assume that it will be easier and useful to create our own smart contract because it will not so difficult and we don't need much functionality atm. Also we can make setup and integration with data access smart contract and Subsqiud more precisely.

keeganquigley commented on applications/staex-iod.md:
```diff
@@ -0,0 +1,326 @@
+
+##### Data access prolongation
+
+For example data cost 1 token per 1KB. If device saved to IPFS 1KB of data it will stop data distribution. If user doesn't want such data access prevents it should put more tokens on smart contract on data request. When provisioner almost done with 1KB it will try to get more tokens from smart contract and in case of success continue to distribute data.
```
Which token are you referring to here? Are you saying it would cost 1 KSM for 1KB?

lavrd replied:
I think it can be any token like DOT/KSM/SDN/..., but for this phase of the project I assume it will be SBY.

keeganquigley commented:
Thanks for the application @lavrd and for being willing to switch to ink! contracts. The [decentralization fund](https://futures.web3.foundation/) you mentioned is a newly launched initiative aimed at kickstarting teams that wish to contribute to the Polkadot/Kusama ecosystem and eventually become self-sustainable. 

One of the main differences is that the futures fund is much broader in scope, and can fund other categories such as marketing, business development, etc. while the W3F grants program is purely for research and technical development. It's completely up to you whether you want to continue to pursue this grant or apply for the fund. Perhaps that is an alternative route if the committee doesn't decide to approve this application. 

In the meantime, I left some suggestions for you to consider.

lavrd commented:
> Thanks for the application @lavrd and for being willing to switch to ink! contracts. The [decentralization fund](https://futures.web3.foundation/) you mentioned is a newly launched initiative aimed at kickstarting teams that wish to contribute to the Polkadot/Kusama ecosystem and eventually become self-sustainable.
> 
> One of the main differences is that the futures fund is much broader in scope, and can fund other categories such as marketing, business development, etc. while the W3F grants program is purely for research and technical development. It's completely up to you whether you want to continue to pursue this grant or apply for the fund. Perhaps that is an alternative route if the committee doesn't decide to approve this application.
> 
> In the meantime, I left some suggestions for you to consider.

Thanks for the clear explanation!

lavrd commented:
@Noc2 @keeganquigley Hi! We see that there is "change requested" label on PR, do we need to clarify or improve something?

lavrd commented:
https://github.com/staex-mcc/staex-iod - our repository for this project.
We have started investigation of Substrate and ink! smart contracts and how to interact with them from Rust code.

semuelle commented:
> Hi! We see that there is "change requested" label on PR, do we need to clarify or improve something?

Hey @lavrd. Sorry for the radio silence. Most of the evaluator team is still out of office, so it might take a couple more days before we get a chance to review again.

keeganquigley reviewed (changes requested):
Sorry for the delay @lavrd I have the following comments:

- There is no **payment address** listed on the application. Could you take a look at the [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md#name-of-your-project) and update based on your preference? If you want cash instead of crypto, you can mark it as FIAT and send banking details separately.
- It's great that you can collect revenue through paying customers, but since many of these components won't make sense to be re-used by others, I wonder if it is the best fit for the grants program. For example, you mention that the entire purpose of the access control layer (M3 & M4) is for restricting usage to paying customers. I'm not sure we'd be willing to fund these parts since they are specific to the product. They also don't seem to relate to substrate/Polkadot tech other than the fact that they would be written in Rust.
- Could you integrate more technical details into M1 & M2? For example, how would the IPFS application work? What does "data exchange" mean and now would it work with the database?

lavrd commented:
@keeganquigley regarding you second question:

Staex itself doesn't collect any revenue or pay to IoT device owners or to users which request data from devices.

The goal of the platform to give IoT device owners possibility to share their data from their devices. So to not share this data for everyone, we need smart contracts (which will be implemented in ink!) to restrict accessing data and open access to particular users. So ACL will be implemented in a following way: provisioner will encrypt data before saving it into IPFS. So only particular user can decrypt data using secret shared key.

M4 more about improving it, as store data is not free for IoT device owner, we need to implement a feature to store this data particular time and delete after expiring or prolong storing this data on user request if user can't download it in time or need it for future.

keeganquigley reviewed (changes requested):
Ah thanks for the clarification @lavrd in this case, perhaps you could incorporate these details into the milestones themselves, along with what the main functions of the ink! contracts would consist of. Since these need to be concrete deliverables that we can evaluate when the time comes. Thanks!

lavrd commented:
@keeganquigley Hi, we have updated our application:

1. We split project details by 4 pats, each part is a separate milestone, so we assume it will be easier for you to understand scope of each milestone.
2. We added more technical details.
3. Payment details were sent.
4. We added additional tech details, they can be expand in file.
5. Also we have updated milestones table, maybe now scope is more clear.

semuelle reviewed (commented):
Thanks for the application, @lavrd. I think Staex looks well thought-out. However, I don't think it's a particularly good fit for the grants program, as the overlap between IOT device owners, data researchers and Polkadot users/developers is fairly small and the focus of the program is on open source projects that are useful to the whole ecosystem. Therefore, I would recommend that you apply at the [Decentralized Futures Program](https://futures.web3.foundation/) instead.

That being said, I will mark your application as ready for review and give the other committee members opportunity to comment and disagree. :)

takahser reviewed (changes requested):
@lavrd I have a couple of questions:

- Did you think of initial markets and specific IoT devices that you'd want to target as well as who would be interested in buying the data?
- What's your approach in finding partnerships and potential customers?
- Also, for UI-based deliverables we'd usually ask for wireframes.
- As for DIDs, have you considered using existing solutions, e.g. [Kilt](https://kilt.io/)?

keeganquigley commented:
Hi @lavrd I have to inform you that unfortunately the committee has unanimously decided not to move forward with this proposal. Thanks again for all your time and efforts in answering our questions. That being said, this proposal may be a better fit for the [Decentralized Futures Program](https://futures.web3.foundation/), as my colleague previously mentioned. Feel free to submit the proposal there instead, and we wish you the best of luck in finding funding for the project.
